### ISSUE TITLE
How can a LEMS file with standard imported content be validated?

### ISSUE THREAD

**Issue by jonrkarr** [NeuroML/pyNeuroML#104]
There appears to be a standard library of content that can be imported.
```xml
<Include file="Cells.xml"/>
<Include file="Networks.xml"/>
<Include file="Simulation.xml"/>
```

How can files that import such content be read and validated? 

The following fails because pynml doesn't know where to find these files, although the `run*` methods must resolve these files.
```python
from pyneuroml import pynml
model = pynml.read_lems_file('LEMS_NML2_Ex5_DetCell.xml', include_includes=True, fail_on_missing_includes=True, debug=False)
```

**Comment by sanjayankur31** [comment]
This calls pylems to read the file and construct the model:

https://github.com/NeuroML/pyNeuroML/blob/development/pyneuroml/pynml.py#L676

In pylems, one has to call `include_file()` or `add_include_directory` first to tell it what the extra files/directories to include/search are. At the moment, the pynml method does not do this:

https://github.com/LEMS/pylems/blob/development/lems/model/model.py#L211

https://github.com/LEMS/pylems/blob/development/lems/model/model.py#L221 

To fix this we need to expose these to the pynml method too. So we can:

- make these standard include files/include search paths included by default
- add arguments to the `read_lems_file` function that allow the user to specify other files + directories to include

@pgleeson what do you think?

**Comment by jonrkarr** [comment]
I worked out how to do this with `lems.model.model.Model.add_include_directory` and the core files in the jar located at `pyneuroml.pynml.get_path_to_jnml_jar()`.

One question: is `NeuroML2CoreTypes/` the only path in the jar file that needs to be automatically included?

Simplifying this would likely be helpful to others. One way to do it would be to unpack the NeuroML2CoreTypes subdirectory of the jar file when pyNeuroML is installed. This can be added to the setup script as a post install command. The jar subdirectory could be unpacked to the installation directory for pyNeuroML. This installation directory can be managed with `pkg_resources.resources_filename`. Then this subdirectory of the installation directory can be used just before https://github.com/NeuroML/pyNeuroML/blob/development/pyneuroml/pynml.py#L676.

**Comment by sanjayankur31** [comment]
Yes, all the LEMS definitions are in the `NeuroML2CoreTypes` folder. I don't think any of the NeuroML tools apart from the jar include these, so yes, shipping them with PyNeuroML is probably the simplest thing to do. We'll look into it.

For reference, all the LEMS definitions for the NeuroML standard are stored here:

https://github.com/NeuroML/NeuroML2/tree/master/NeuroML2CoreTypes

The schema itself is here:

https://github.com/NeuroML/NeuroML2/tree/master/Schemas/NeuroML2

**Comment by sanjayankur31** [comment]
Ideally, the way to validate a LEMS NeuroML2 file is to use use jnml with `-norun`. The `read_lems_file` function will read any LEMS file and is not limited to NeuroML files. See https://github.com/NeuroML/pyNeuroML/pull/108

**Comment by jonrkarr** [comment]
What is the difference between `lems.model.model.Model().import_from_file()` and `jnml -norun`? What does the latter do that the former doesn't?

**Comment by jonrkarr** [comment]
I see that `lems.model.model.Model().import_from_file()` provides much less validation. For example, no error is thrown for missing targets of simulations.

**Comment by sanjayankur31** [comment]
Yeh, the PyLEMS functions are meant for any LEMS files but they will be validated against the LEMS schema. `jnml` validates against the NeuroML schema (which uses LEMS to define component types too) and includes some extra tests:

https://github.com/NeuroML/org.neuroml.model/blob/development/src/main/java/org/neuroml/model/util/NeuroML2Validator.java

**Comment by pgleeson** [comment]
> What is the difference between `lems.model.model.Model().import_from_file()` and `jnml -norun`? What does the latter do that the former doesn't?

@jonrkarr It's important to point out that jlems and pylems are independent implementations of the LEMS language, and that jlems is much more complete, and as such is the reference implementation. Basically, if jlems can read it, understand it and run it, it's "official" LEMS.  

There is a scheme file for LEMS, which the NeuroML lems definitions are valid against, but jlems is much more forgiving, and will allow specification of files in any order as long as they are complete. 

PyLems is a pure python implemnetation of the language, but there are relatively recent additions to LEMS which it does not yet support, like Tunnels (and so no gap junctions can be simulated with it).
